File: DSAinJava/DynamicPrograming.java/CoinChange.java
import java.util.Arrays;

/**
 * leetcode problem no 322
 * leetcode solution lin(https://leetcode.com/problems/coin-change/)
 */

 class coinChange {

    public static int[] c; // global coin array
    public static int[] dp;
    public int f(int amount ){
        if(amount < 0) return Integer.MAX_VALUE; //if ans is less then 0  then it's not possible
        if(amount == 0 ) return 0; // if ans is 0 , give nothing

        if(dp[amount ] != -1) return dp[amount]; 

        int ans= Integer.MAX_VALUE;
        int  n= c.length;
        for(int i =0; i<n; i++){
            if(amount >= c[i]){
                ans = Math.min(ans , f(amount - c[i]));
            }
        }

        // if ans is still Integer.MAX_VALUE , then this amount is not  possible
        if(ans == Integer.MAX_VALUE){

            return dp[amount] =Integer.MAX_VALUE;
        }
        return dp[amount] =1 + ans;
    }
    public int coinChange(int[] coins, int amount) {

        c =coins;
        dp = new int[10005];
        Arrays.fill(dp,-1);
        int ans = f(amount);
        return (ans == Integer.MAX_VALUE) ? -1 : ans;


     
        
        
    }
}

File: DSAjavascript/DynamicPrograming.js/coinChange.js
/**
 * leetcode problem no 322
 * leetcode solution lin(https://leetcode.com/problems/coin-change/)
 */
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    let dp = new Array(amount+1).fill(0);
    //from dp[0] to dp[amount]
    //init dp[0] = 0
    //calculate dp[1] to dp[amount]
    for(let i=1; i<=amount; i++) {
        let min = Number.POSITIVE_INFINITY
        for(j=0; j<coins.length; j++) {
            if(i-coins[j]>=0) {
                min = Math.min(min, 1+dp[i-coins[j]])
            }
        }
        dp[i] = min
    }
    return dp[amount] !=  Number.POSITIVE_INFINITY ? dp[amount] : -1
};
